user/mgetty/voice/man/man1/rmdtopvf.1
.\" .IX pvf
.TH "rmdtopvf" "1" "1.4" "rmdtopvf" "PVF tools"
.SH "NAME"
\fBrmdtopvf\fR \- convert from RMD to PVF format.
.SH "SYNOPSIS"
rmdtopvf [options] [<rmdfile> [<pvffile>]]
.SH "DESCRIPTION"
\fIrmdtopvf\fR converts from the rmd (raw modem data) format to the pvf (portable voice format) format.
.SH "OPTIONS"
rmdtopvf can be called in the pipe mode:

	cat infile.rmd | rmdtopvf [optons] > outfile.pvf

or with two file names:

	rmdtopvf [options] infile.rmd outfile.pvf

Valid options are:

\-h     brief help message

\-L     list of supported raw modem data formats

\-a     output pvf ascii format

\-b     output pvf binary format (default)

\-8     output 8 bit samples

\-16    output 16 bit samples

\-32    output 32 bit samples (default)


Usually there is no need for any of them.


.SH "FORMATS"
Supported raw modem data formats:

 \- Digi           4 (G.711U) and 5 (G.711A)

 \- Elsa           2, 3 and 4 bit Rockwell ADPCM

 \- ISDN4Linux     2, 3 and 4 bit ZyXEL ADPCM

 \- Multitech 2834 4 bit IMA ADPCM

 \- Rockwell       2, 3 and 4 bit Rockwell ADPCM

 \- Rockwell       8 bit Rockwell PCM

 \- V253modem      8 bit PCM

 \- UMC            4 (G.721 ADPCM)

 \- US Robotics    1 and 4 (GSM and G.721 ADPCM

 \- ZyXEL 1496     2, 3 and 4 bit ZyXEL ADPCM

 \- ZyXEL 2864     2, 3 and 4 bit ZyXEL ADPCM

 \- ZyXEL Omni 56K 4 bit Digispeech ADPCM (?)

.SH "SEE ALSO"
autopvf, lintopvf, voctopvf, basictopvf, wavtopvf, pvftoau, pvftowav, pvftolin, pvftovoc, pvftobasic, and pvftormd.
.SH "EXAMPLES"
Play a rmd file using play from the SOX package:

	rmdtopvf rmdfile | pvftowav | play \-t .wav \-


Play a rmd file by dumping it directly to /dev/audio:

	rmdtopvf rmdfile | pvftoau > /dev/audio

.SH "AUTHOR"
Mikkel L. Ellertson <mikkel@infinity\-ltd.com>


freeswan/CHANGES
New and notable in 1.97:

Sanity check added for at least one of AH or ESP more for checking if
"make *config" has even been run.

Fixed 2.2 local IKE fragmentation blackhole.  Still won't work if
iptraf or another pcap app is running.

MCR added ikeping(8).

Applied DHR's tunnel patch to streamline IKE/specialSA processing.

Added local and remote IKE bypass tests to the UML test suite.

UML KLIPS %trap* tests now check for upbound pf_key messages.

An ICMP ICMP_DEST_UNREACH, ICMP_PKT_FILTERED is now sent upon %reject.

Added a UML input file with first and last packets being outside the 
covered range to "sandwich" the packets under test.

The default _updown script has been cleaned up slightly, reducing
internal duplication that arose accidentally.

When doing --add on an RSA-signature conn, ipsec_auto now tells Pluto
about the keys before, not after, telling it about the conn.  The main
consequence of this is that if something is wrong with one of the keys,
the conn itself does *not* get loaded.  This should reduce confusion.

The rp_filter checks during KLIPS startup have been fixed to be a bit less
paranoid; rp_filter is thought to cause problems only on the physical
interfaces, not on the ipsecN interfaces. 

The configuration process has been fixed to be a bit more paranoid.  Some
users have had mysterious compile failures with complaints involving names
like MD5_CTX; they seem to have resulted from kernel configuration changes
somehow getting into .config but not into include/linux/autoconf.h (which
is what actually controls most of the kernel compile).  The post-config
checking now verifies not only that .config has the right stuff in it, but
that autoconf.h does too.  (We still don't see *how* this discrepancy can
arise, but this will at least catch it when it does.)

The details of linking the KLIPS source into the kernel are now under
control of a configuration variable in Makefile.inc, to make it easier
for distribution builders to change.

The top-level Makefile now includes cleverness which assigns a date-based
version code if a build is done from a raw CVS checkout.

The library now includes a new function, ttodatav(), like ttodata() but
in some cases it gives more verbose error reports.  Also, there are now
a couple of functions which distill an RSA public key to a "key ID" for
use in messages, a crypto-quality pseudo-random-number generator, and a
function that converts a single address into a (singleton) subnet.

Barf's secret/key-censoring utilities have been tidied up so that what
they emit matches the key-ID library functions. 

The ipsec command now supplies a couple more environment variables to
things it calls, and this has reduced the number of things which need to
be edited at build time to know about configuration variables. 

The configuration reader now allows parameter names to begin with "_";
these are strictly reserved for internal development purposes and will
not in general be documented.


As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.96:

Compatibility with the 2.0.xx kernels (under Red Hat 5.2) is broken.
This was accidental, and we have some hopes of fixing it, although it
won't last forever even if we do.  See README.

Our "IKE Implementation Issues" IETF draft has been added to our docs.

Newhostkey now has a mandatory --output "option" specifying a filename
(although the magic filename "-" means "standard output"), and sets
restrictive permissions on newly-created files to try to keep private keys
truly private. 

Pluto now gracefully handles a couple of odd error cases (missing private
keys, network errors at just the wrong time in negotiations) which used to
cause it to die, provoking an automatic restart of the IPsec subsystem. 

Makefile.inc now has more detailed control over where manpages go, to ease
integration into distributions.  The name of the destination directory for
RPMs is also now centralized there.

Rsasigkey now checks for certain obscure internal failures, which can
happen if --random is used to supply a source of random bits that aren't
really very random, and complains accordingly.

Barf's eroute table, formerly sorted by source, is now sorted by destination.
This is probably a bit clearer for hosts with lots of eroutes.

Use of the %reject target in a shunt eroute now, as long planned, actually
causes an ICMP rejection message (Destination Unreachable:  Communication
Administratively Prohibited). 

KLIPS now declares its module license to be "Dual BSD/GPL".  (It is 
dual because the libdes and zlib code is BSD licensed, while KLIPS itself 
is GPL).  Whether this is exactly right is not yet clear, but at least it
stops the latest module loaders from whining.

The UML test stuff continues to grow and add new test cases.

As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.95:

The last remnants of the "%hold" bug, which broke 1.93 and 1.94, have
(we think) been dealt with.

Minor bugs in the RPM-building stuff have been fixed, and it has been
updated to deal with Red Hat's abrupt removal of the --buildroot option
from rpm.

Pluto now supports the ability to delete an instance of a conn, rather
than just the whole conn (for cases, notably OE, where a single connection
description can be involved in multiple connections).  There is no support
for this in ipsec_auto yet.

Pluto's routine (non-debug) logging is now somewhat terser, and some of
the more boring messages are produced only if --debug-lifecycle is turned
on (this is a temporary hack, it's not really debugging output).

Pluto's innards have been reorganized somewhat, in preparation for making
DNS lookups asynchronous.  There should be no user-visible effects.

Pluto now ignores IPv6 link-local interface addresses; we think that they
are never relevant, and trying to bind to them causes difficulties.  This

user/mgetty/voice/pvftools/rmdtopvf.c
/*
 * rmdtopvf.c
 *
 * rmdtopvf converts from the rmd (raw modem data) format to the pvf
 * (portable voice format) format.
 *
 * $Id: rmdtopvf.c,v 1.19 2005/03/13 17:27:48 gert Exp $
 *
 */

#include "../include/voice.h"

char *program_name;

static void usage (void)
     {
     fprintf(stderr, "\n%s %s\n\n", program_name, vgetty_version);
     fprintf(stderr, "usage:\n");
     fprintf(stderr, "\t%s [options] [<rmdfile> [<pvffile>]]\n", program_name);
     fprintf(stderr, "\noptions:\n");
     fprintf(stderr, "\t-h     this help message\n");
     fprintf(stderr, "\t-L     list of supported raw modem data formats\n");
     fprintf(stderr, "\t-a     output pvf ascii format\n");
     fprintf(stderr, "\t-b     output pvf binary format (default)\n");
     fprintf(stderr, "\t-8     output 8 bit samples\n");
     fprintf(stderr, "\t-16    output 16 bit samples\n");
     fprintf(stderr, "\t-32    output 32 bit samples (default)\n\n");
     exit(ERROR);
     }

static void supported_formats (void)
     {
     fprintf(stderr, "\n%s %s\n\n", program_name, vgetty_version);
     fprintf(stderr, "supported raw modem data formats:\n\n");
     fprintf(stderr, " - Digi RAS       G.711u/A PCM\n");
     fprintf(stderr, " - Elsa           2, 3 and 4 bit Rockwell ADPCM\n");
     fprintf(stderr, " - ISDN4Linux     2, 3 and 4 bit ZyXEL ADPCM,\n");
     fprintf(stderr, "                  G.711u/A PCM\n");
     fprintf(stderr, " - Lucent         8/16 bit linear PCM, G.711u/A PCM,\n");
     fprintf(stderr, "                  4 bit IMA ADPCM\n");
     fprintf(stderr, " - Multitech 2834 4 bit IMA ADPCM\n");
     fprintf(stderr, " - Multitech 5634 4 bit IMA ADPCM\n");
     fprintf(stderr, " - Rockwell       2, 3 and 4 bit Rockwell ADPCM,\n");
     fprintf(stderr, "                  8 bit Rockwell PCM\n");
     fprintf(stderr, " - UMC            G.721 ADPCM\n");
     fprintf(stderr, " - US Robotics    USR-GSM, G.721 ADPCM\n");
     fprintf(stderr, " - V.253          2 and 4 bit Rockwell ADPCM,\n");
     fprintf(stderr, "                  4 bit IMA ADPCM\n");
     fprintf(stderr, "                  8 bit linear signed/unsigned PCM\n");
     fprintf(stderr, "                  16 bit linear signed PCM Intel Order\n");
     fprintf(stderr, "                  G.711u/A PCM\n");
     fprintf(stderr, " - ZyXEL 1496     2, 3 and 4 bit ZyXEL ADPCM\n");
     fprintf(stderr, " - ZyXEL 2864     2, 3 and 4 bit ZyXEL ADPCM\n");
     fprintf(stderr, "                  8 bit Rockwell PCM\n\n");
     fprintf(stderr, " - ZyXEL Omni 56K 4 bit Digispeech ADPCM (?)\n");
     exit(ERROR);
     }

int main (int argc, char *argv[])
     {
     int option;
     FILE *fd_in = stdin;
     FILE *fd_out = stdout;
     char *name_in = "stdin";
     char *name_out = "stdout";
     rmd_header header_in;
     pvf_header header_out = init_pvf_header;
     char *modem_type;
     int compression;

     check_system();
     program_name = argv[0];

     while ((option = getopt(argc, argv, "abhL12368")) != EOF)
          {

          switch (option)
               {
               case 'a':
                    header_out.ascii = TRUE;
                    break;
               case 'b':
                    header_out.ascii = FALSE;
                    break;
               case 'L':
                    supported_formats();
               case '8':
                    header_out.nbits = 8;
                    break;
               case '1':
               case '6':
                    header_out.nbits = 16;
                    break;
               case '3':
               case '2':
                    header_out.nbits = 32;
                    break;
               default:
                    usage();
               };

          };

     if (optind < argc)
          {
          name_in = argv[optind];

          if ((fd_in = fopen(name_in, "r")) == NULL)
               {
               fprintf(stderr, "%s: Could not open file %s\n", program_name,
                name_in);
               exit(ERROR);
               };

          optind++;
          };

     if (read_rmd_header(fd_in, &header_in) != OK)
          exit(ERROR);

     header_out.speed = ntohs(header_in.speed);
     modem_type = header_in.voice_modem_type;
     compression = ntohs(header_in.compression);

     if (header_out.speed == 0)
          {

          if ((strcmp(modem_type, "ZyXEL 1496") == 0) ||
           (strcmp(modem_type, "ZyXEL 2864") == 0) ||
           (strcmp(modem_type, "ZyXEL Omni 56K") == 0))
               header_out.speed = 9600;

          if ((strcmp(modem_type, "Rockwell") == 0) ||
           (strcmp(modem_type, "Elsa") == 0))
               header_out.speed = 7200;

          if (strcmp(modem_type, "Lucent") == 0)
	         header_out.speed = 8000;

          if (strcmp(modem_type, "ISDN4Linux") == 0)
	         header_out.speed = 8000;
 
         }

     if (optind < argc)
          {
          name_out = argv[optind];

          if ((fd_out = fopen(name_out, "w")) == NULL)
               {
               fprintf(stderr, "%s: Could not open file %s\n", program_name,
                name_out);
               exit(FAIL);
               };

          };

     if (write_pvf_header(fd_out, &header_out) != OK)
          {
          fclose(fd_out);

          if (fd_out != stdout)
               unlink(name_out);

          exit(ERROR);
          }

     if (strcmp(modem_type, "Digi RAS") == 0 && compression == 4)
          {

          if (ulawtopvf(fd_in, fd_out, &header_out) == OK)
               exit(OK);

          };

     if (strcmp(modem_type, "Digi RAS") == 0 && compression == 5)
          {

	  if (alawtopvf(fd_in, fd_out, &header_out) == OK)
	       exit(OK);

          };

     if ((strcmp(modem_type, "Elsa") == 0) && ((compression == 2) ||
      (compression == 3) || (compression == 4)))
          {

          if (rockwelltopvf(fd_in, fd_out, compression, &header_out) == OK)
               exit(OK);

          };

     if ((strcmp(modem_type, "ISDN4Linux") == 0) && ((compression == 2) ||
      (compression == 3) || (compression == 4)))
          {

          if (zyxeltopvf(fd_in, fd_out, compression, &header_out) == OK)
               exit(OK);

          };

     if (strcmp(modem_type, "ISDN4Linux") == 0 && compression == 5)
          {

	  if (alawtopvf(fd_in, fd_out, &header_out) == OK)
	       exit(OK);

          };

     if (strcmp(modem_type, "ISDN4Linux") == 0 && compression == 6)
          {

          if (ulawtopvf(fd_in, fd_out, &header_out) == OK)
               exit(OK);

          };

     if (strcmp(modem_type, "Lucent") == 0 && compression == 1)
          {

          if (lintopvf(fd_in, fd_out, &header_out, 0, 8, 0) == OK)
               exit(OK);

          };

     if (strcmp(modem_type, "Lucent") == 0 && compression == 2)
          {

          if (lintopvf(fd_in, fd_out, &header_out, 0, 16, 0) == OK)
               exit(OK);

          };

     if (strcmp(modem_type, "Lucent") == 0 && compression == 3)
          {

	  if (alawtopvf(fd_in, fd_out, &header_out) == OK)
	       exit(OK);

          };

     if (strcmp(modem_type, "Lucent") == 0 && compression == 4)
          {

          if (ulawtopvf(fd_in, fd_out, &header_out) == OK)
               exit(OK);

          };

     if (strcmp(modem_type, "Lucent") == 0 && compression == 5)
          {

          if (imaadpcmtopvf(fd_in, fd_out, &header_out) == OK)
               exit(OK);

          };

     if (strcmp(modem_type, "Lucent") == 0 && compression == 6)
          {

	/*          if (g728topvf(fd_in, fd_out, compression, &header_in) == OK)
		    exit(OK); */
	fprintf(stderr, "%s: G.728 compression is not yet supported!\n",
                 program_name);

          };

     if (((strcmp(modem_type, "Multitech2834") == 0) || (strcmp(modem_type, "Multitech5634") == 0)) && ((compression == 4) || compression == 132))
          {

          if (imaadpcmtopvf(fd_in, fd_out, &header_out) == OK)
               exit(OK);

          };

     if ((strcmp(modem_type, "Rockwell") == 0) && ((compression == 2) ||
      (compression == 3) || (compression == 4)))
          {

          if (rockwelltopvf(fd_in, fd_out, compression, &header_out) == OK)
               exit(OK);

          };

     if (strcmp(modem_type, "Rockwell") == 0 && compression == 8)
          {

	  if (rockwellpcmtopvf(fd_in, fd_out, compression, &header_out) == OK)
               exit(OK);

          };

     if (strcmp(modem_type, "UMC") == 0 && compression == 4)
          {

          if (usrtopvf(fd_in, fd_out, compression, &header_out) == OK)
               exit(OK);

          };

     if ((strcmp(modem_type, "US Robotics") == 0) && ((compression == 1) ||
      (compression == 4)))
          {

          if (usrtopvf(fd_in, fd_out, compression, &header_out) == OK)
               exit(OK);

          };

     if (strcmp(modem_type, "V253modem")==0)
	{
          switch(compression)
          {
          case 0:
          case 1:
          case 8:
            if (lintopvf(fd_in, fd_out, &header_out, 0, 0, 0) == OK)
               exit(OK);
            else exit(FAIL);
          case 9:
            if (lintopvf(fd_in, fd_out, &header_out, 1, 0, 0) == OK)   /* signed linear */
               exit(OK);
            else exit(FAIL);
          case 12:
            /* signed linear PCM 16-bit Inter bit order */
            if (lintopvf(fd_in, fd_out, &header_out, 1, 16, 1) == OK)
               exit(OK);
            else exit(FAIL);
          case 13:
            /* unsigned linear PCM 16-bit Inter bit order */
            if (lintopvf(fd_in, fd_out, &header_out, 0, 16, 1) == OK)
               exit(OK);
            else exit(FAIL);
          case 2:
          case 4:
            if (rockwelltopvf(fd_in, fd_out, compression, &header_out) == OK)   /* the &Elsa compatible formats */
               exit(OK);
            else exit(FAIL);
          case 5:
            if (imaadpcmtopvf(fd_in, fd_out, &header_out) == OK)
               exit(OK);
            else exit(FAIL);
          case 6:
          case 10:
            if (ulawtopvf(fd_in, fd_out, &header_out) == OK)
                exit(OK);
            else exit(FAIL);
          case 7:
          case 11:
            if (alawtopvf(fd_in, fd_out, &header_out) == OK)
                exit(OK);
            else exit(FAIL);

          default:
            fprintf(stderr, "%s: Unsupported compression method (%s/%d)\n",
              program_name, modem_type, compression);
            exit(FAIL);
          };
        };

     if (((strcmp(modem_type, "ZyXEL 1496") == 0) ||
      (strcmp(modem_type, "ZyXEL 2864") == 0)) &&
      ((compression == 2) || (compression == 3) || (compression == 30) ||
      (compression == 4)))
          {

          if (zyxeltopvf(fd_in, fd_out, compression, &header_out) == OK)
               exit(OK);

          };

     if (strcmp(modem_type, "ZyXEL 2864") == 0 && compression == 81)
          {
          if (rockwellpcmtopvf(fd_in, fd_out, compression, &header_out) == OK)
               exit(OK);
          };

     if (strcmp(modem_type, "ZyXEL Omni 56K") == 0 && compression == 4)
          {

          if (zo56ktopvf(fd_in, fd_out, &header_out) == OK)
               exit(OK);

          };

     fclose(fd_out);

     if (fd_out != stdout)
          unlink(name_out);

     fprintf(stderr, "%s: Unsupported compression method (%s/%d)\n",
      program_name, modem_type, compression);
     exit(FAIL);
     }
